File: golang/daily21/m01/d02.go
// 每日一题 20210102
// 239. 滑动窗口最大值
// Link: https://leetcode-cn.com/problems/sliding-window-maximum/
package y2021m01

func maxSlidingWindow(nums []int, k int) []int {
	if k < 2 {
		return nums
	}
	var q []int
	push := func(i int) {
		for len(q) > 0 && nums[i] >= nums[q[len(q)-1]] {
			q = q[:len(q)-1]
		}
		q = append(q, i)
	}

	for i := 0; i < k; i++ {
		push(i)
	}
	n := len(nums)
	ret := make([]int, 1, n-k+1)
	ret[0] = nums[q[0]]
	for i := k; i < n; i++ {
		push(i)
		for q[0] <= i-k {
			q = q[1:]
		}
		ret = append(ret, nums[q[0]])
	}
	return ret
}

File: typescript/daily21/m07/d02.ts
import { runAllTest } from "../../utils/test"
/**
 * 
 * 1833. 雪糕的最大数量
 * @link https://leetcode-cn.com/problems/maximum-ice-cream-bars/
 */
function maxIceCream(costs: number[], coins: number): number {
  const cs = [...costs].sort((a, b) => a - b);
  let ret = 0;
  for (const c of cs) {
    if (coins < c) {
      return ret;
    }
    coins -= c;
    ret += 1;
  }
  return ret;
};

runAllTest("20210702", maxIceCream, [
  { args: [[1, 3, 2, 4, 1], 7], res: 4 },
  { args: [[10, 6, 8, 7, 7, 8], 5], res: 0 },
  { args: [[1, 6, 3, 1, 2, 5], 20], res: 6 },
  { args: [[10, 2, 10, 10, 10, 10, 8, 2, 7, 8], 25], res: 4 },
]);